File: web/lib/voice.ts
/**
 * Voice service integration functions
 * Provides speech-to-text and text-to-speech capabilities
 */

const API_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/api/v1';

/**
 * Speech-to-text conversion
 * @param audioFile - Audio file to convert to text
 * @param language - Language code (default: en-US)
 * @returns Promise resolving to transcription result
 */
export const speechToText = async (
  audioFile: File,
  language: string = 'en-US'
): Promise<{ text: string; language: string; confidence: number }> => {
  try {
    const formData = new FormData();
    formData.append('audio_file', audioFile);
    formData.append('language', language);

    const token = typeof window !== 'undefined' ? localStorage.getItem('token') : null;
    const response = await fetch(`${API_URL}/voice/speech-to-text`, {
      method: 'POST',
      body: formData,
      headers: token ? { Authorization: `Bearer ${token}` } : {},
    });

    if (!response.ok) {
      throw new Error(`Speech-to-text request failed: ${response.status}`);
    }

    const json = await response.json();
    return json.data;
  } catch (error) {
    console.error('Speech-to-text error:', error);
    throw new Error('Failed to convert speech to text');
  }
};

/**
 * Text-to-speech conversion
 * @param text - Text to convert to speech
 * @param voice - Voice identifier (default: default)
 * @param speed - Speech speed (0.5 to 2.0, default: 1.0)
 * @param pitch - Speech pitch (0.5 to 2.0, default: 1.0)
 * @returns Promise resolving to audio blob
 */
export const textToSpeech = async (
  text: string,
  voice: string = 'default',
  speed: number = 1.0,
  pitch: number = 1.0
): Promise<Blob> => {
  try {
    const formData = new FormData();
    formData.append('text', text);
    formData.append('voice', voice);
    formData.append('speed', speed.toString());
    formData.append('pitch', pitch.toString());

    const token = typeof window !== 'undefined' ? localStorage.getItem('token') : null;
    const response = await fetch(`${API_URL}/voice/text-to-speech`, {
      method: 'POST',
      body: formData,
      headers: token ? { Authorization: `Bearer ${token}` } : {},
    });

    if (!response.ok) {
      throw new Error(`Text-to-speech request failed: ${response.status}`);
    }

    return response.blob();
  } catch (error) {
    console.error('Text-to-speech error:', error);
    throw new Error('Failed to convert text to speech');
  }
};

File: backend/app/api/endpoints/voice.py
"""
Voice processing endpoints for NovaMind AI
Provides speech-to-text and text-to-speech capabilities
"""

from fastapi import APIRouter, HTTPException, Depends, File, UploadFile, Form
from fastapi.responses import Response
from typing import Optional
import logging
from app.core.speech_service import speech_to_text, text_to_speech
from app.api.deps import get_current_active_user
from app.models.user import User

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/voice", tags=["voice"])

@router.post("/speech-to-text")
async def speech_to_text_endpoint(
    audio_file: UploadFile = File(...),
    language: str = Form("en-US"),
    current_user: User = Depends(get_current_active_user)
):
    """
    Convert speech audio to text

    Args:
        audio_file: Uploaded audio file
        language: Language code for recognition (default: en-US)
        current_user: Currently authenticated user

    Returns:
        Transcription results
    """
    try:
        # Read audio file content
        audio_data = await audio_file.read()

        # Process speech to text
        result = await speech_to_text(audio_data, language)

        logger.info(f"User {current_user.id} processed speech-to-text: {len(audio_data)} bytes")

        return {
            "success": True,
            "data": result
        }
    except Exception as e:
        logger.error(f"Error in speech-to-text endpoint: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Speech-to-text processing failed: {str(e)}")

@router.post("/text-to-speech")
async def text_to_speech_endpoint(
    text: str = Form(...),
    voice: str = Form("default"),
    speed: float = Form(1.0),
    pitch: float = Form(1.0),
    current_user: User = Depends(get_current_active_user)
):
    """
    Convert text to speech audio

    Args:
        text: Text to convert to speech
        voice: Voice identifier to use
        speed: Speech speed (0.5 to 2.0)
        pitch: Speech pitch (0.5 to 2.0)
        current_user: Currently authenticated user

    Returns:
        Audio file response
    """
    try:
        # Validate input
        if not text.strip():
            raise HTTPException(status_code=400, detail="Text cannot be empty")

        if len(text) > 5000:  # Reasonable limit
            raise HTTPException(status_code=400, detail="Text too long (max 5000 characters)")

        # Process text to speech
        audio_bytes = await text_to_speech(text, voice, speed, pitch)

        logger.info(f"User {current_user.id} processed text-to-speech: {len(text)} characters")

        # Return audio file
        return Response(
            content=audio_bytes,
            media_type="audio/wav",
            headers={
                "Content-Disposition": "attachment; filename=speech.wav"
            }
        )
    except Exception as e:
        logger.error(f"Error in text-to-speech endpoint: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Text-to-speech processing failed: {str(e)}")